pandas/core/ops/__init__.py

```diff
@@ -293,6 +294,84 @@ def fill_binop(left, right, fill_value):
 # Dispatch logic
 
 
+def operate_blockwise(left, right, array_op):
+    assert right._indexed_same(left)
+
+    def get_same_shape_values(
+        lblk: "Block", rblk: "Block", left_ea: bool, right_ea: bool
+    ) -> Tuple[ArrayLike, ArrayLike]:
+        """
+        Slice lblk.values to align with rblk.  Squeeze if we have EAs.
+        """
+        lvals = lblk.values
+        rvals = rblk.values
+
+        # TODO(EA2D): with 2D EAs pnly this first clause would be needed
+        if not (left_ea or right_ea):
+            lvals = lvals[rblk.mgr_locs.indexer, :]
+            assert lvals.shape == rvals.shape, (lvals.shape, rvals.shape)
+        elif left_ea and right_ea:
+            assert lvals.shape == rvals.shape, (lvals.shape, rvals.shape)
+        elif right_ea:
+            # lvals are 2D, rvals are 1D
+            lvals = lvals[rblk.mgr_locs.indexer, :]
+            assert lvals.shape[0] == 1, lvals.shape
+            lvals = lvals[0, :]
+        else:
+            # lvals are 1D, rvals are 2D
+            assert rvals.shape[0] == 1, rvals.shape
+            rvals = rvals[0, :]
```

Comment:
Can you help me understand what cases gets here? This is from something like adding a 2D float64 block to an IntegerArray in a DataFrame?

```python
pd.DataFrame(np.random.random(size=(10, 2)), columns=['a', 'b']) + pd.DataFrame({"a": pd.array(list(range(10)))})
```

Reply:
> This is from something like adding a 2D float64 block to an IntegerArray in a DataFrame?

Yes, but in the example there should only be one column in the first DataFrame:

```
left = pd.DataFrame(np.random.random(size=(10, 1)), columns=['a'])
right = pd.DataFrame({"a": pd.array(list(range(10)))})

lvals = left._data.blocks[0].values
rvals = right._data.blocks[0].values

>>> lvals.shape
(1, 10)
>>> rvals.shape
(10,)

```